File: backend/domain/src/transforms/transform_port.rs
use serde::{Deserialize, Serialize};
use crate::db::db_transform::TransformPortId;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransformPort {
    pub port_id: TransformPortId,
    pub name: String,
    pub direction: String, // "input" | "output"
    pub port_order: i32,
    pub description: Option<String>,
}

File: frontend/src/domain/Transform/TransformPort.ts

export interface TransformPort {
  port_id: number;
  name: string;
  direction: "input" | "output";
  port_order: number;
  description?: string;
  // Program = main signal (fails closed if unwired); Sidechain = control/
  // detector signal (unwired always resolves to silence). See
  // agents/transforms.md's ABI contract section.
  kind: "program" | "sidechain";
  // Single = exactly one edge may target this port; Many = N edges sum into
  // it (opt-in, e.g. multiple mic feeds into one "mix" port).
  cardinality: "single" | "many";
}